go/types/value_store.go

```diff
@@ -353,8 +353,8 @@ func (lvs *ValueStore) checkChunksInCache(v Value, readValues bool) Hints {
 		}
 
 		targetType := getTargetType(reachable)
-		if !entry.Type().Equals(targetType) {
-			d.Panic("Value to write contains ref %s, which points to a value of a different type: %+v != %+v", reachable.TargetHash(), entry.Type(), targetType)
+		if entry.Type().Kind() != ValueKind && targetType.Kind() != ValueKind && !entry.Type().Equals(targetType) {
```

Comment:
I think we need the old behavior here. Which is that if the type is more specific than the information we have, we're forced to load the value and check its actual type.